ntoskrnl/se/token.c

```diff
@@ -135,34 +135,138 @@ SepDeleteTokenLock(
 
     /* Lock the tokens */
     SepAcquireTokenLockShared(FirstToken);
     SepAcquireTokenLockShared(SecondToken);
 
-    /* FIXME: Check if every SID that is present in either token is also present in the other one */
+    /* Check if every SID that is present in either token is also present in the other one */
+    for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->UserAndGroupCount; FirstTokenIndex++)
+    {
+        for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->UserAndGroupCount; SecondTokenIndex++)
+        {
+            if (!RtlEqualSid(FirstToken->UserAndGroups[FirstTokenIndex].Sid,
+                             SecondToken->UserAndGroups[SecondTokenIndex].Sid))
+            {
+                DPRINT1("SepCompareTokens(): SIDs are not equal!\n");
+                goto Quit;
+            }
+
+            /* Are the attributes from either SID from tokens equal? */
+            if (FirstToken->UserAndGroups[FirstTokenIndex].Attributes !=
+                SecondToken->UserAndGroups[SecondTokenIndex].Attributes)
+            {
+                DPRINT1("SepCompareTokens(): SID attributes are not equal!\n");
+                goto Quit;
+            }
+        }
+    }
 
     Restricted = SeTokenIsRestricted(FirstToken);
     if (Restricted == SeTokenIsRestricted(SecondToken))
     {
         if (Restricted)
         {
-            /* FIXME: Check if every SID that is restricted in either token is also restricted in the other one */
+            /* Check if every SID that is restricted in either token is also restricted in the other one */
+            for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->RestrictedSidCount; FirstTokenIndex++)
+            {
+                for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->RestrictedSidCount; SecondTokenIndex++)
+                {
+                    if (!RtlEqualSid(FirstToken->RestrictedSids[FirstTokenIndex].Sid,
+                                     SecondToken->RestrictedSids[SecondTokenIndex].Sid))
+                    {
+                        DPRINT1("SepCompareTokens(): Restricted SIDs are not equal!\n");
+                        goto Quit;
+                    }
+                }
+            }
+
+            /* Check if every privilege that is present in either token is also present in the other one */
+            for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->PrivilegeCount; FirstTokenIndex++)
+            {
+                for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->PrivilegeCount; SecondTokenIndex++)
+                {
+                    /* Are the privileges from either token equal? */
+                    if (FirstToken->Privileges[FirstTokenIndex].Luid.LowPart !=
+                        SecondToken->Privileges[SecondTokenIndex].Luid.LowPart)
+                    {
+                        DPRINT1("SepCompareTokens(): Privileges are not equal!\n");
+                        goto Quit;
```

Comment:
continue

Reply:
By using your suggestion, the code just simply jumps to the second loop with the first one never firing up to loop over the first array... 

Furthermore according to `SepComparePrivilegeAndAttributeArrays`'s behaviour if privileges and/or attributes are not equal it doesn't continue looping the arrays but instead it bails out and `NtCompareTokens` deems the tokens as not equivalent (same goes for SIDs and their attributes).